### ISSUE TITLE
GPU running in thermostatic mode?

### ISSUE THREAD

**Issue by csecht** [Ricks-Lab/gpu-utils#50]
For a while one of my two RX 570s ran in what appeared to be a thermostatic mode while the other ran as expected from amdgpu-pac settings. 

This occurred after upgrading to the amdgpu-pro-19.30-934563-ubuntu-18.04 package (release date November 5th 2019) with the command *./amdgpu-pro-install -y --opencl=pal,legacy --headless*, then rebooting. I have my systemd set to run *amdgpu-pac* bash scripts at startup that set PWM fan speeds and other card parameters for einstein@home crunching. While one card ran its PAC settings as expected, the other card ignored its prescribed PWM setting (45%) and instead ran variable fan speeds which kept the GPU around 74 C. Over the course of a day or so, fan speeds ran between 36% and 43%, in 1% increments, as GPU load and room temperature varied. I have never seen that happen before.  I tried to reproduce this behavior on the other card by setting its fan PWM to default (-1), but that did not trigger the thermostatic mode. Unfortunately, when I rebooted the system again, all fan PWM settings returned to normal.   Has anyone else has experienced this?
 
I've no idea what happened, but it would be great to have the option in *amdgpu-pac* to take advantage of this (mysterious) AMDGPU feature and set fan PWM to "thermostatic".

**Comment by csecht** [comment]
Follow up: After a system shutdown and restart, for some reason the autostart PAC bash script for Card2 didn't take (but did for Card1), so the card ran at it's default AMDGPU driver settings.  As before, its fan PWM ran with fluctuating fan speeds to maintain a relatively constant temperature (mid-70s C). When I tried to set a P-state mask for Card2, *amdgpu-pac* returned a bash I/O error and didn't echo the changes.  After a subsequent system reboot (done for unrelated reasons), the autostart PAC scripts successfully set run parameters for both cards and both responded normally to additional *amdgpu-pac* edits (i.e., with no fluctuating fan speeds). 
Today and initially, both  instances of "thermostatic" PWM affected Card2.

**Comment by csecht** [comment]
Update: While in "thermostatic mode", fan PWM does not go below 33%; when that minimum is reached,  the GPU temperature falls below the target 74-75 C range as the room temperature cools.  I'll probably need to wait until the summer months to see how high fan speed goes to maintain the target GPU temperature.

**Comment by csecht** [comment]
update: After a prior update to AMDGPU 19.5 from 19.3, thermostatic mode, stopped working. Following a subsequent Ubuntu system update today for which I also did an AMDGPU uninstall and reinstall of OpenCL in headless mode, thermostatic mode is working again, but only for Card 1 (previously only worked with Card 2). Truly baffling. I'm currently running Ubuntu 18.04.4, kernel 5.3.0-40-generic, and AMDGPU 19.50-967956 with AMD RX 570 cards.

**Comment by csecht** [comment]
I found a partial explanation, but have uncovered another mystery. The PAC script that I have set to run at startup by systemd lists hwmon2 for Card1 whereas it is actually hwmon3, so the sh command to manually set the PWM was ignored and the card booted up in a default "dynamic" setting, which provides GPU thermostasis.  Apparently the hwmon number for a device (either card) can change following kernel or AMDGPU updates. From *amdgpu-ls*, this default (thermostatic) fan setting is listed as: Fan PWM Mode: [2, 'Dynamic'], (where the pwm1_enable file contains a value of 2).
However, when using *amdgpu-pac* to set the Fan PWM of that card to "auto" (setting of -1), the fan controller then switches out of a thermostatic mode and into a mode that constantly alternates between a fan speed of 26% and 49% or thereabouts, and the temperature fluctuates with room temperature. BUT *amdgpu-ls* STILL REPORTS the setting as: Fan PWM Mode: [2, 'Dynamic']. So in a default startup scenario, a PWM mode of 2 is thermostatic, but when set by an *amdgpu-pac* script, a PWM mode of 2 is not. In the former instance fan speed varies incrementally, while in the later it varies only between two apparently pre-set speeds. What's going on?

**Comment by csecht** [comment]
As an ad hoc fix, users can ensure a thermostatic mode by commenting out the lines in the PAC sh scripts that write to files pwm1_enable and pwm1 in the hwmon directories.

**Comment by Ricks-Lab** [comment]
Let me dig into some more detailed documentation I found recently. Also, if modifications are needed, I will make them in the v3.0 branch.

**Comment by Ricks-Lab** [comment]
There is mention at the bottom of [comment38 ](https://bugzilla.kernel.org/show_bug.cgi?id=201539#c38) about restarting services.  Not sure if this is related to your observations.

**Comment by csecht** [comment]
That's an interesting discussion thread that seems like it should be relevant, but am not sure how.  It's not clear to me how the PWM auto setting (pwm1_enable = 2) is working for them.  Restarting my startup systemd service didn't help. There's something magical happening at system restart (or reboot?) that changes the behavior of the card's default fan and temperature control. I also don't understand what a pwm1_enable value of 0 does.  I wonder in which hwmgr file the thermostatic setpoint of 75 C is defined. 

Comment 45 from your link does explain why my startup PAC script stopped manually setting fan PWM (because the hwmon# changes): "The "HWMON" variable is there to determine which actual hardware monitor is being used because it can change whenever you boot. One time it could be hwmon1, the next time hwmon3, etc. So you can't hard-code it as you're doing. You have to use the $HWMON variable to set FILE_PWM, FILE_FANMODE, and FILE_TEMP." 
Using that kind of $HWMON variable would help. I'll try to work it into the script.

**Comment by Ricks-Lab** [comment]
I am not sure what I can do differently in pac.  Currently, GPUs seem to be in dynamic mode by default.  I need to set to manual in order to specify a pwm value.  To reset a GPU, I just set mode back to dynamic.  None of my cards have fans, so I can not experiment with them.  I suggest you go to the hwmon directory and manually write to the `pwm` and `pwm1_enable` sensors to get a better understanding of what is happening.

For your script, would an option to ls that gives output something like this for each GPU be helpful?
```
card_num:hwmon_path
```

**Comment by csecht** [comment]
Good idea.  I just need to play around more with pwm and pwm1_enable and try to figure out how dynamic mode can change from flipping between two fan speeds to changing incrementally, and back again. 

A card’s hwmon path is already defined in the second section of ls output, so I don’t think a separate option is needed. (or am I missing something?)  On some restarts the card number also changes; only a card’s PCIe number stays constant. I need to think about how to generalize those variables in a startup PAC script.

**Comment by Ricks-Lab** [comment]
I thought a special option for ls would make it easier to parse in your script. Maybe a ‘--raw’ option for parsable output would be useful. 

Also, it would be good to verify what mode zero does. With the name being ‘no fan control’, I suspect pwm signal is off, resulting in full fan speed. It would be good to verify.

**Comment by csecht** [comment]
Verified:  "0" in pwm1_enable sets full fan speed

**Comment by csecht** [comment]
... and resetting back to "1" (manual) leaves it at full speed; "2" (auto) resets fan pwm to oscillating between ~26% and ~49%.

**Comment by csecht** [comment]
Just to confirm: 
1) When I edited my systemd startup PAC scripts for both cards to comment out the lines that manually set fan PWM, then, upon reboot, fans on both cards ran in a default/auto thermostatic mode. I've not yet figured out how to enter this mode variant without a reboot/restart.

2) When I edited a startup PAC script to replace the hardcoded hwmon number with a variable for setting fan PWM, then it worked fine (code snipet for Card1 below).
```
set -x
# Assign correct hwmon number, which is subject to change each boot.
HWMON=$(ls /sys/class/drm/card1/device/hwmon/)
# Fan PWM Old:  44 New:  47 Min:  0 Max:  100
sudo sh -c "echo '1' >  /sys/class/drm/card2/device/hwmon/$HWMON/pwm1_enable"
sudo sh -c "echo '119' >  /sys/class/drm/card2/device/hwmon/$HWMON/pwm1"
```
Such a modification allows startup PAC scripts that manually set fan PWM to continue working across shutdowns, reboots, upgrades, etc.

**Comment by Ricks-Lab** [comment]
Doesn’t seem like there is anything I can do in pac at this point. Perhaps if we can find out what is different after reboot, we can find a solution. In the meantime, I will implement a message saying reboot is required when returning to dynamic mode. I will let you know when I have something to test in v3.0.

**Comment by csecht** [comment]
I wonder whether others see the type of oscillating fan pwm (between ~26% and ~49%) when I or pac write “2” to pwm1_enable after startup. There are a few variables that might affect it: kernel, amdgpu, PWM control, GPU BIOS?.

**Comment by csecht** [comment]
I just updated my AMDGPU drivers on my other E@H host so now both hosts have the same Linux kernal. It runs an AMD CPU and two RX 460s (in contrast to the Intel CPU and RX 570 host I've discussed above). Fan PWM behavior is very similar between the two hosts. The only difference is that the RX 460 machine, while in the reboot-default-auto mode, does incrementally adjust fan speed, but only between 18% and 25% (usually stays at 20%), and doesn't maintain GPU thermostasis.
